### ISSUE TITLE
BlkDiagMatrix Apply/Solve

### ISSUE THREAD

**Issue by garrettwrong** [ComputationalCryoEM/ASPIRE-Python#277]
Both `apply` and `solve` are implemented a little strange. Ostensibly apply is just doing a matrix-vector or matrix-matrix multiply over ragged arrays where the RHS is provided as a flattened vector or matrix instead of BlkDiagMatrix. (Pretty much everything up to the operation is same for solve).  We use some legacy conversion code that can be deprecated `src/aspire/utils/cell.py`to do this. That code is not used anywhere else iirc.

Three approaches (of many):

1) Just manage an index into the RHS, slicing. Then we loop through computing block @ slice. This takes a few lines vs page of confusing code.

2)  Expand RHS to a BlkDiagMatrix (pack as block diags), or sequence of BlkDiagMatrix for matrix case, then use matmul, contract BlkDiagMatrix (unpack diags from block). This uses existing matmul, within the current limitations of both class instances being square and matching sizes.

3) Extend BlkDiagMatrix to non square matrices.  This would mean Apply becomes matmul.  The non square-ness feature could be done wholesale, or just for matmul to support this.

This is cleanup/optimization, so we can discuss what makes more sense for the future.  We do use these methods, so I think leaning them out is a good idea. I just found the code I was considering to use in the unit test reference was much much simpler than the actual code...

```
        # Manually compute one matrix vector product                                                                                                              
        ind = 0
        res = np.empty_like(coeffm)
        for b, blk in enumerate(self.blk_a):
            col = self.blk_a.partition[b,1]
            res[ind: ind + col, :] = blk @ coeffm[ind: ind + col, :]
            ind += col
```

**Comment by janden** [comment]
I don't think it makes sense to think of the RHS in `apply`/`solve` as a `BlkDiagMatrix`. Yes, the effect is the same for these functions, but not for other operations, like “applying” that RHS to another vector or matrix. The natural object represented by the RHS is a *partitioned* or *block* matrix (of which a block diagonal matrix is a specific type). Schematically, we have
```
[[A 0 0]    [[D]    [[A * D]
 [0 B 0]  *  [E]  =  [B * E]
 [0 0 C]]    [F]]    [C * F]]
```
which gives the same result, block-per-block, as
```
[[A 0 0]    [[D 0 0]    [[A * D 0 0]
 [0 B 0]  *  [0 E 0]  =  [0 B * E 0]
 [0 0 C]]    [0 0 F]]    [0 0 C * F]]
```
but the complete objects are not the same.

I do agree that the `cell` code mostly obfuscates things here. From what I remember, it's there to simulate certain Matlab behavior that we used to achieve some speedups (for some reason, it was faster to split up an arry into a cell matrix instead of doing the indexing manually). So I would vote for approach 1. If these types of matrices crop up more often, we may want to consider implementing some sort of `BlkMatrix` and subclass `BlkDiagMatrix` from that, but I think the marginal benefit right now is quite low.

**Comment by garrettwrong** [comment]
Right, for 2, if we wrapped apply to convert the RHS input to blk diag, we would have to wrap it on the way back from the matmul to extract the expected vector/matrix. The loop with slicing (1) is pretty simple, and we'd need that sort of loop as part of the unpack in 2 anyway, so I guess 2 doesn't make much sense...

That is interesting about the `cell` code.  I can't say I am a Matlab performance wizard.  I really wouldn't have guessed the generically typed cell would be faster than array arithmetic.  I wonder why that was so.

I agree this isn't something we need to optimize yet, just noting. Thanks

**Comment by janden** [comment]
Well with 2 we're still forcing the data into a structure that's not appropriate for it, which could cause problems down the road. I think 1 is the way to go for now.

IIRC the issue was more the partitioning that was the issue (with `mat2cell`, `cell2mat`, and `num2cell`). Not sure if it was faster or just ended up being more idiomatic Matlab.
